pandas/core/reshape/reshape.py

```diff
@@ -643,8 +643,15 @@ def _convert_level_number(level_num, columns):
             levs.append(np.take(lev, level_codes))
         tuples = list(zip(*levs))
         unique_groups = [key for key, _ in itertools.groupby(tuples)]
-        new_names = this.columns.names[:-1]
-        new_columns = MultiIndex.from_tuples(unique_groups, names=new_names)
+        new_columns = MultiIndex.from_arrays(
+            [
+                Index(new_level, dtype=level.dtype)
+                if None not in new_level
```

Comment:
do we have any tests that hit the None case?

Reply:
Yes, specifically `test_stack_nan_in_multiindex_columns`.

`None` is a tricky case---it is allowed for some Index types (like `CategoricalIndex`) but not others (like `Int64Index`). I am just avoiding levels with `None` entirely.

I've also looked at the way `None` is handled elsewhere, and it's not totally consistent:

```
>>> MultiIndex.from_arrays([[1, 2, None]]).levels[0]
Int64Index([1, 2], dtype='int64')
```

But:

```
>>> Index([1, 2, None])
Index([1, 2, None], dtype='object')
```

Perhaps in the future it would be better if `Index` accepted `None` values, even when a dtype is specified. Then `Index([1, 2, None], dtype='int64')` would return `Int64Index([1, 2], dtype='int64')`. Then we wouldn't need such a condition here. Thoughts?

Reply:
hmm interesting, ok can you open an issue specifically showing the MI vs Index cases. I agree we should do something about this. ok for here on this PR.